src/coord/src/catalog.rs

```diff
@@ -1254,10 +1263,62 @@ impl Catalog {
 
     /// Returns a mapping that indicates all indices that are available for
     /// each item in the catalog.
-    pub fn indexes(&self) -> &HashMap<GlobalId, Vec<Vec<ScalarExpr>>> {
+    pub fn indexes(&self) -> &HashMap<GlobalId, Vec<(GlobalId, Vec<ScalarExpr>)>> {
         &self.indexes
     }
 
+    /// Finds the nearest indexes that can satisfy the views or sources whose
+    /// identifiers are listed in `ids`.
+    ///
+    /// Returns the identifiers of all discovered indexes, along with a boolean
+    /// indicating whether the set of indexes is complete. If incomplete, then
+    /// one of the provided identifiers transitively depends on an
+    /// unmaterialized source.
+    pub fn nearest_indexes(&self, ids: &[GlobalId]) -> (Vec<GlobalId>, bool) {
+        fn inner(
+            catalog: &Catalog,
+            id: GlobalId,
+            indexes: &mut Vec<GlobalId>,
+            complete: &mut bool,
+        ) {
+            // If an index exists for `id`, record it in the output set and stop
+            // searching.
+            if let Some((index_id, _)) = catalog.indexes[&id].first() {
+                indexes.push(*index_id);
+                return;
+            }
+
+            match catalog.get_by_id(&id).item() {
+                view @ CatalogItem::View(_) => {
+                    // Unmaterialized view. Recursively search its dependencies.
+                    for id in view.uses() {
+                        inner(catalog, id, indexes, complete)
+                    }
+                }
+                CatalogItem::Source(_) => {
+                    // Unmaterialized source. Record that we are missing at
+                    // least one index.
+                    *complete = false;
+                }
+                CatalogItem::Table(_) => {
+                    unreachable!("tables always have at least one index");
```

Comment:
What if someone manually drops the index?
```
materialize=> CREATE TABLE t (a int, b text NOT NULL);
CREATE TABLE
materialize=> SHOW INDEXES FROM t;
    Source_or_view    |             Key_name             | Column_name | Expression | Null | Seq_in_index
----------------------+----------------------------------+-------------+------------+------+--------------
 materialize.public.t | materialize.public.t_primary_idx | a           |            | t    |            1
 materialize.public.t | materialize.public.t_primary_idx | b           |            | f    |            2
(2 rows)

materialize=> drop index materialize.public.t_primary_idx;
DROP INDEX
materialize=> SHOW INDEXES FROM t;
 Source_or_view | Key_name | Column_name | Expression | Null | Seq_in_index
----------------+----------+-------------+------------+------+--------------
(0 rows)
```

Reply:
Yep, totally! (See #3905.) I think we just need to disable dropping the default index on tables, since it's really unusual that you could drop an index on a table and accidentally delete all the data in it.

Reply:
That makes sense! In the meantime, this isn't actually `unreachable!()` though. Should we return an error instead and put a todo? Or are we okay with this panic?

Reply:
I'm ok with it for just a moment since a) tables are experimental, and b) I'll have a PR out by end of day! Thanks for raising, though! I won't forget!